wallet-new/src/Cardano/Wallet/Kernel/Transactions.hs

```diff
@@ -163,17 +176,28 @@ newTransaction ActiveWallet{..} genChangeAddr signAddress options accountId paye
 -- when we estimate the fees and later create a transaction, the coin selection
 -- will always yield the same value, making the process externally-predicatable.
 newtype PayMonad a = PayMonad {
-      buildPayment :: ReaderT (NonEmpty (Address, Coin)) IO a
-    } deriving (Functor, Applicative, Monad, MonadIO, MonadReader (NonEmpty (Address, Coin)))
+      buildPayment :: ReaderT (ByteString, IORef Word16) IO a
+    } deriving ( Functor
+               , Applicative
+               , Monad
+               , MonadIO
+               , MonadReader (ByteString, IORef Word16)
+               )
 
 -- | \"Invalid\" 'MonadRandom' instance for 'PayMonad' which generates
--- randomness using the hash of 'NonEmpty (Address, Coin)' as fixed seed.
+-- randomness using the hash of 'NonEmpty (Address, Coin)' as fixed seed,
+-- plus an internal counter used to shift the bits of such hash.
+-- This ensures that the coin selection algorithm runs in a random environment
+-- which is yet deterministically reproduceable by feeding the same set of
+-- payees.
 instance MonadRandom PayMonad where
-    getRandomBytes _ =
-        ask >>= return . ByteArray.convert
-                       . encodeUtf8 @Text @ByteString
-                       . sformat build
-                       . hash
+    getRandomBytes _ = do
```

Comment:
Although it is better, I still feel somewhat uncomfortable with this code. We have IO available. Couldn't we call into a proper random number generator, but using your `PayMonad` state to give it a seed? That way we'd have predictability but we wouldn't be trying to roll our own PRNG.